Sources/AsyncHTTPClient/RedirectState.swift

```diff
@@ -0,0 +1,140 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import struct Foundation.URL
+import NIOHTTP1
+
+struct RedirectState {
+    /// number of redirects we are allowed to follow.
+    private var count: Int
+
+    /// All visited URLs.
+    private var visited: [String]
+
+    /// if true, `redirect(to:)` will throw an error if a cycle is detected.
+    private let allowCycles: Bool
+}
+
+extension RedirectState {
+    /// Creates a `RedirectState` from a configuration.
+    /// Returns nil if the user disallowed redirects,
+    /// otherwise an instance of `RedirectState` which respects the user defined settings.
+    init?(
+        _ configuration: HTTPClient.Configuration.RedirectConfiguration.Configuration,
+        initialURL: String
+    ) {
+        switch configuration {
+        case .disallow:
+            return nil
+        case .follow(let maxRedirects, let allowCycles):
+            self.init(count: maxRedirects, visited: [initialURL], allowCycles: allowCycles)
+        }
+    }
+}
+
+extension RedirectState {
+    /// Call this method when you are about to do a redirect to the given `redirectURL`.
+    /// This method records that URL into `self`.
+    /// - Parameter redirectURL: the new URL to redirect the request to
+    /// - Throws: if it reaches the redirect limit or detects a redirect cycle if and `allowCycles` is false
+    mutating func redirect(to redirectURL: String) throws {
+        guard self.count > 0 else {
+            throw HTTPClientError.redirectLimitReached
+        }
+
+        if !allowCycles && self.visited.contains(redirectURL) == true {
+            throw HTTPClientError.redirectCycleDetected
+        }
+
+        self.count -= 1
```

Comment:
Should we store `count` or just derive it from `self.visited`?

Reply:
`self.visited.count` counts up, `self.count` counts down until it hits zero where we know we reached the limit.
Because we need to know when we reach the limit, `self.visited` wouldn't be enough and we would still need to store the limit. Nevertheless, I think it's clearer to only count up and compare `self.visited.count` agains the constant limit to deduplicate counting.